### PR TITLE
Tools: Extract parameter default values from an ArduPilot .bin log file

### PR DESCRIPTION
**Pull request by amilcarlucas**
Contains unittests with 96% coverage

@peterbarker how should I integrate the unittests with the autotest suite?

### CODE REVIEW

**Comment by amilcarlucas** [comment]
To run coverage do:
```
cd Tools/autotest/param_metadata
python3 -m coverage run -m unittest extract_param_defaults_unittest.py
python3 -m coverage html
firefox htmlcov/extract_param_defaults_py.html
```

**Comment by khancyr** [comment]
we don't gather coverage for python code for now. Adding it should quite simple, generate the coverage report and put it in the upload directory for coveralls. Then it should appears online.

**Comment by amilcarlucas** [comment]
Thanks. My question is how to integrate the unit tests? Coverage information is just a cherry on top. Nice but not important now.

**Comment by amilcarlucas** [comment]
I am trying to solve the problem of managing (without MP) multiple sets of parameters for multiple vehicles.
I want to store only the parameters that change relative to the defaults of a particular vehicle in a particular configuration in a particular version.

I do not think that generating default .param files and distributing them is a good Idea. I do not distribute them on https://discuss.ardupilot.org/t/how-to-methodically-tune-almost-any-multicopter-using-arducopter-4-4-x/110842 for the reasons you mention.

But I do need to get then in a machine readable format so that I can automate their management. This is the reason for this script.

**Comment by andyp1per** [comment]
I agree with @IamPete1 about providing defaults files - but this does not do that - it just makes it easier to get at them which seems ok to me

**Comment by amilcarlucas** [comment]
All checks pass

File: Tools/scripts/extract_param_defaults_unittest.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
+#!/usr/bin/python3
```
Comment (tridge) [inline Tools/scripts/extract_param_defaults_unittest.py:1]: could be in Tools/scripts/tests

**Review by tridge — CHANGES_REQUESTED** [review]

File: .github/workflows/test_extract_params_default.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  push:
+    paths:
+      - 'Tools/scripts/extract_params_default.py'
+      - 'Tools/scripts/extract_params_default_unittest.py'
+
+jobs:
+  build:
+    runs-on: ubuntu-22.04
+    container:
+      image: ardupilot/ardupilot-dev-clang:v0.1.3
```
Comment (khancyr) [inline .github/workflows/test_extract_params_default.yml:13]: dev-base

File: .github/workflows/test_extract_params_default.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      options: --user 1001
```
Comment (khancyr) [inline .github/workflows/test_extract_params_default.yml:14]: needed ?

File: .github/workflows/test_extract_params_default.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+      - name: Set up Python
+        uses: actions/setup-python@v2
```
Comment (khancyr) [inline .github/workflows/test_extract_params_default.yml:19]: we don't need the python setup nor Install dependencies as this is already done by the base image

**Review by khancyr — CHANGES_REQUESTED** [review]

File: Tools/scripts/extract_param_defaults_unittest.py (hunk 1 above)
Comment (amilcarlucas) [reply Tools/scripts/extract_param_defaults_unittest.py]: I did that, but the linter does not like it :(

File: .github/workflows/test_extract_params_default.yml (hunk 2 above)
Comment (amilcarlucas) [reply .github/workflows/test_extract_params_default.yml]: done

File: .github/workflows/test_extract_params_default.yml (hunk 3 above)
Comment (amilcarlucas) [reply .github/workflows/test_extract_params_default.yml]: done, I think.

File: .github/workflows/test_extract_params_default.yml (hunk 4 above)
Comment (amilcarlucas) [reply .github/workflows/test_extract_params_default.yml]: Done, I think.

**Review by tridge — CHANGES_REQUESTED** [review]

File: Tools/scripts/extract_param_defaults_unittest.py (hunk 1 above)
Comment (amilcarlucas) [reply Tools/scripts/extract_param_defaults_unittest.py]: Done.

**Comment by amilcarlucas** [comment]
Tests pass.

**Review by tridge — APPROVED** [review]

**Pull request merged**